File: src/algorithms/matrices/construction/toeplitz-matrix/__tests__/ToeplitzMatrix_test.java
// javac ToeplitzMatrix.java ToeplitzMatrix_test.java && java -ea ToeplitzMatrix_test

public class ToeplitzMatrix_test {

    public static void main(String[] args) {
        testCanonicalToeplitzExample();
        testNonToeplitz2x2();
        testSingleElementMatrix();
        testSingleRowMatrix();
        testSingleColumnMatrix();
        testAllSameElements();
        testValid2x2Toeplitz();
        testInvalid2x2NonToeplitz();
        testFirstRowMismatch();
        testLastDiagonalBroken();
        System.out.println("All tests passed!");
    }

    static void testCanonicalToeplitzExample() {
        int[][] matrix = {{1, 2, 3, 4}, {5, 1, 2, 3}, {9, 5, 1, 2}};
        assert ToeplitzMatrix.toeplitzMatrix(matrix) == true;
    }

    static void testNonToeplitz2x2() {
        int[][] matrix = {{1, 2}, {2, 2}};
        assert ToeplitzMatrix.toeplitzMatrix(matrix) == false;
    }

    static void testSingleElementMatrix() {
        int[][] matrix = {{42}};
        assert ToeplitzMatrix.toeplitzMatrix(matrix) == true;
    }

    static void testSingleRowMatrix() {
        int[][] matrix = {{1, 2, 3, 4}};
        assert ToeplitzMatrix.toeplitzMatrix(matrix) == true;
    }

    static void testSingleColumnMatrix() {
        int[][] matrix = {{1}, {2}, {3}};
        assert ToeplitzMatrix.toeplitzMatrix(matrix) == true;
    }

    static void testAllSameElements() {
        int[][] matrix = {{7, 7, 7}, {7, 7, 7}, {7, 7, 7}};
        assert ToeplitzMatrix.toeplitzMatrix(matrix) == true;
    }

    static void testValid2x2Toeplitz() {
        int[][] matrix = {{1, 2}, {3, 1}};
        assert ToeplitzMatrix.toeplitzMatrix(matrix) == true;
    }

    static void testInvalid2x2NonToeplitz() {
        int[][] matrix = {{5, 3}, {3, 4}};
        assert ToeplitzMatrix.toeplitzMatrix(matrix) == false;
    }

    static void testFirstRowMismatch() {
        int[][] matrix = {{1, 2, 3}, {4, 2, 2}, {7, 4, 2}};
        assert ToeplitzMatrix.toeplitzMatrix(matrix) == false;
    }

    static void testLastDiagonalBroken() {
        int[][] matrix = {{1, 2, 3}, {4, 1, 2}, {7, 4, 9}};
        assert ToeplitzMatrix.toeplitzMatrix(matrix) == false;
    }
}

File: src/algorithms/matrices/construction/toeplitz-matrix/__tests__/toeplitz-matrix_test.rs
include!("../sources/toeplitz-matrix.rs");

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_canonical_toeplitz_example() {
        let matrix = vec![vec![1, 2, 3, 4], vec![5, 1, 2, 3], vec![9, 5, 1, 2]];
        assert_eq!(toeplitz_matrix(&matrix), true);
    }

    #[test]
    fn test_non_toeplitz_2x2() {
        let matrix = vec![vec![1, 2], vec![2, 2]];
        assert_eq!(toeplitz_matrix(&matrix), false);
    }

    #[test]
    fn test_single_element_matrix() {
        let matrix = vec![vec![42]];
        assert_eq!(toeplitz_matrix(&matrix), true);
    }

    #[test]
    fn test_single_row_matrix() {
        let matrix = vec![vec![1, 2, 3, 4]];
        assert_eq!(toeplitz_matrix(&matrix), true);
    }

    #[test]
    fn test_single_column_matrix() {
        let matrix = vec![vec![1], vec![2], vec![3]];
        assert_eq!(toeplitz_matrix(&matrix), true);
    }

    #[test]
    fn test_all_same_elements() {
        let matrix = vec![vec![7, 7, 7], vec![7, 7, 7], vec![7, 7, 7]];
        assert_eq!(toeplitz_matrix(&matrix), true);
    }

    #[test]
    fn test_valid_2x2_toeplitz() {
        let matrix = vec![vec![1, 2], vec![3, 1]];
        assert_eq!(toeplitz_matrix(&matrix), true);
    }

    #[test]
    fn test_invalid_2x2_non_toeplitz() {
        let matrix = vec![vec![5, 3], vec![3, 4]];
        assert_eq!(toeplitz_matrix(&matrix), false);
    }

    #[test]
    fn test_first_row_mismatch() {
        let matrix = vec![vec![1, 2, 3], vec![4, 2, 2], vec![7, 4, 2]];
        assert_eq!(toeplitz_matrix(&matrix), false);
    }

    #[test]
    fn test_last_diagonal_broken() {
        let matrix = vec![vec![1, 2, 3], vec![4, 1, 2], vec![7, 4, 9]];
        assert_eq!(toeplitz_matrix(&matrix), false);
    }
}
